### PR TITLE
8263191: Consolidate ThreadInVMfromJavaNoAsyncException and ThreadBlockInVMWithDeadlockCheck with existing wrappers

### PR DESCRIPTION
**Pull request by pchilano**
Hi,

Please review the following small patch. ThreadInVMfromJavaNoAsyncException is exactly the same as ThreadInVMfromJava except that in the destructor we avoid checking for asynchronous exceptions. Similarly ThreadBlockInVMWithDeadlockCheck is equal to ThreadBlockInVM except that in the destructor we might need to release a Mutex in case we need to stop for a safepoint or handshake.
We can consolidate these two wrappers with the existing ones to avoid code duplication and minimize the number of transition wrappers.

Tested in mach5 tiers 1-3.

Thanks,
Patricio

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263191](https://bugs.openjdk.java.net/browse/JDK-8263191): Consolidate ThreadInVMfromJavaNoAsyncException and ThreadBlockInVMWithDeadlockCheck with existing wrappers

### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to e99ca599a056c3dba15292ef48a735cdd4c93fa1
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to e99ca599a056c3dba15292ef48a735cdd4c93fa1
 * [Daniel D. Daugherty](https://openjdk.java.net/census#dcubed) (@dcubed-ojdk - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2880/head:pull/2880`
`$ git checkout pull/2880`

### CODE REVIEW

**Review by coleenp — APPROVED** [review]
Looks good!

**Review by dholmes-ora — APPROVED** [review]
Looks good Patricio!

Thanks,
David

File: src/hotspot/share/runtime/interfaceSupport.inline.hpp [hunk 1]
```diff
@@ -150,17 +150,19 @@ class ThreadInVMForHandshake : public ThreadStateTransition {
 };
 
 class ThreadInVMfromJava : public ThreadStateTransition {
+  bool _check_asyncs;
  public:
-  ThreadInVMfromJava(JavaThread* thread) : ThreadStateTransition(thread) {
+  ThreadInVMfromJava(JavaThread* thread, bool check_async = true) : ThreadStateTransition(thread), _check_asyncs(check_async) {
```
Comment (dcubed-ojdk) [inline src/hotspot/share/runtime/interfaceSupport.inline.hpp:155]: Field is `_check_asyncs` and parameter is `check_async`.
Either add an `s` to the parameter or delete the `s` from the field.

File: src/hotspot/share/runtime/interfaceSupport.inline.hpp [hunk 2]
```diff
@@ -222,93 +224,43 @@ class ThreadToNativeFromVM : public ThreadStateTransition {
...
-    assert(_in_flight_mutex_addr != NULL, "_in_flight_mutex_addr should have been set on constructor");
-    Mutex* in_flight_mutex = *_in_flight_mutex_addr;
-    if (in_flight_mutex != NULL) {
-      in_flight_mutex->release_for_safepoint();
-      *_in_flight_mutex_addr = NULL;
-    }
-  }
  public:
-  ThreadBlockInVMWithDeadlockCheck(JavaThread* thread, Mutex** in_flight_mutex_addr)
+  ThreadBlockInVM(JavaThread* thread, Mutex** in_flight_mutex_addr = NULL)
```
Comment (dcubed-ojdk) [inline src/hotspot/share/runtime/interfaceSupport.inline.hpp:237]: The new ThreadBlockInVM is not equivalent to the old one.
The old one used `trans(_thread_in_vm, _thread_blocked)` which
asserted that the "from" state is equal to `_thread_in_vm`. This
new code does not do that. The trans() call also resulted in a
`SafepointMechanism::process_if_requested(thread);` call after
we transitioned the thread to `_thread_in_vm`+1 and before we
set the new thread state to `_thread_blocked`.

File: src/hotspot/share/runtime/interfaceSupport.inline.hpp [hunk 3]
```diff
@@ -222,93 +224,43 @@ class ThreadToNativeFromVM : public ThreadStateTransition {
...
+    thread->check_possible_safepoint();
     // Once we are blocked vm expects stack to be walkable
     thread->frame_anchor()->make_walkable(thread);
-
-    // All unsafe states are treated the same by the VMThread
-    // so we can skip the _thread_in_vm_trans state here. Since
-    // we don't read poll, it's enough to order the stores.
-    OrderAccess::storestore();
-
     thread->set_thread_state(_thread_blocked);
```
Comment (dcubed-ojdk) [inline src/hotspot/share/runtime/interfaceSupport.inline.hpp:242]: If you're going to set the thread_state directly here, I think you need
to keep the `OrderAccess::storestore()` call. However, this new ThreadBlockInVM
is also missing a `SafepointMechanism::process_if_requested(thread)` call that
happened before setting the `_thread_blocked` state.

**Review by dcubed-ojdk — CHANGES_REQUESTED** [review]
I think there are non-trivial differences between the old ThreadBlockInVM
and the new ThreadBlockInVM. Maybe I'm wrong, but they do not look
equivalent to me.

**Comment by dcubed-ojdk** [comment]
I think you're missing my point about the equivalence of the old ThreadBlockInVM()
and the new ThreadBlockInVM(). The old version did this in trans()/transition():

```
    // Change to transition state and ensure it is seen by the VM thread.
    thread->set_thread_state_fence((JavaThreadState)(from + 1));

    SafepointMechanism::process_if_requested(thread);
    thread->set_thread_state(to);
```

which means that when we were in (`_thread_blocked`+1) we did a
`process_if_requested(thread)` call before we changed the thread's
state to `_thread_blocked`. In the new code, we just got straight to
`_thread_blocked` without processing any pending request. I think the
only case that matters is that we won't process a pending handshake
request. As you say, a pending safepoint will be allowed to proceed
once we've reached `_thread_blocked`.

It could be that this change in behavior doesn't make any real difference,
but I want to be sure.

**Review by dcubed-ojdk — CHANGES_REQUESTED** [review]
Thanks for making the fixes you've done so far, but I still
have one nagging question...

**Review by dcubed-ojdk — APPROVED** [review]
Patricio and I have bounced this change in behavior back and forth
in Slack. I'm now convinced that it is a benign change in behavior.

**Comment by pchilano** [comment]
Thanks @coleenp, @dholmes-ora and @dcubed-ojdk for the reviews!

**Pull request closed without merging**